File: 2017/d8/sol1.rb
input = File.read("input.txt").split("\n")

registers = Hash.new(0)

max_val = 0

input.each do |inst|
  reg1, action, val, f, reg2, comp, comp_val = inst.split

  cond = false
  case comp
    when '<'
      cond =  registers[reg2] < comp_val.to_i
    when '<='
      cond =  registers[reg2] <= comp_val.to_i
    when '=='
      cond = registers[reg2] == comp_val.to_i
    when '!='
      cond = registers[reg2] != comp_val.to_i
    when '>='
      cond = registers[reg2] >= comp_val.to_i
    when '>'
      cond = registers[reg2] > comp_val.to_i
    else
      raise "wtf: #{inst}"
  end

  if cond
    val = val.to_i
    if action == 'dec'
      val *= -1
    end

    registers[reg1] += val
  end

  max_val = [max_val, registers.values.max || 0].max
end

puts registers.values.max
puts max_val

File: 2017/d8/sol1.ex
registers = %{}

File.read!("input.txt")
  |> String.trim
  |> String.split("\n", trim: true)
  |> Enum.each(fn(x) ->
      [reg1, action, val, z, reg2, comp, comp_val] = String.split(x)

      if registers[reg1] == nil do
        registers[reg1] = 0
      end

      if registers[reg2] == nil do
        registers[reg2] = 0
      end
    end)
  |> IO.puts
